### PR TITLE
IPv6 support, phase 1 (direct connection)

### PR DESCRIPTION
**Pull request by softins**
This implementation of IPv6 on Jamulus is based on #1017 by @jardous, with further enhancements by @softins. It is an alternative to #1455 by @mstolle629.

It has been tested on Linux, Mac and Windows. It is currently running on jamminthejazz, at `privjam.softins.co.uk:22123` (IPv4) and `privjam6.softins.co.uk:22123` (IPv6). There is also a private instance at `privjam.softins.co.uk` and `privjam6.softins.co.uk`, which may be used for testing.

Servers running this version can accept a mix of IPv4 and IPv6 connections.

Server addresses for direct connection can be specified in the following formats:

- `[v6addr]`
- `[v6addr]:port`
- `[v4addr]`
- `[v4addr]:port`
- `v4addr`
- `v4addr:port`
- `hostname`
- `hostname:port`

A literal IPv6 address must always be enclosed in `[ ]`, to avoid ambiguity with a port.

Phase 2 will be the addition of Directory Server support, which will require some
protocol enhancements. Phase 1 is worth adding on its own before Phase 2 is ready.

**Potential ChangeLog entry**

Add support for direct connections using IPv6

**Documentation needed**
- Command line options `-6` and `--enableipv6`
- Address formats as shown above

**Status of this Pull Request**

Working implementation

**What is missing until this pull request can be merged?**

More testing is always good!

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

File: src/socket.cpp
Comment (atsampson) [inline src/socket.cpp]: Missing space after setsockopt (I guess clang-format will pick this up anyway).

File: src/socket.cpp
Comment (atsampson) [inline src/socket.cpp]: The rest of the network code avoids writing `struct` when it's not necessary in C++ (`sockaddr`, `sockaddr_in`); can this and the sizeof below just be `sockaddr_in6`?

File: src/socket.cpp
Comment (atsampson) [inline src/socket.cpp:250]: `CVector<T>` subclasses `std::vector<T>`, so can this just use vecbySendBuf.c_data() to get a const pointer rather than copying the vector?

Comment (softins) [reply src/socket.cpp:250]: Maybe, but I didn't change that part of the code.

File: src/socket.cpp
Comment (softins) [reply src/socket.cpp]: Probably best to review just the differences between the branch point and the final commit.

File: src/client.cpp
Comment (pljones) [inline src/client.cpp:34]: It doesn't need the ugly `N`, generally.

**Comment by pljones** [comment]
Um...  Given IPv4 is always enabled and given that only IPv4 is used to communicate with directory servers (for now), is there any reason not to have IPv6 always enabled?

File: src/client.cpp
Comment (softins) [reply src/client.cpp:34]: Was just following what appeared to be the majority convention. No strong view either way.

**Comment by softins** [comment]
I wondered that. It would fail on a system that didn't have an IPv6 stack. Such as a Linux box with a file containing "blacklist ipv6" in /etc/modprobe.d/ (I did test that on my Pi). We could try creating an AF_INET6 socket first and falling back to AF_INET, bit would then need to propagate the result of that to a global variable that all classes could consult.

**Comment by pljones** [comment]
Well, rather than set `bEnableIPv6` in `main.cpp` based on a command line parameter, set it based on the result of the check, leaving the rest the same?  (More `bIPv6Available`, in other words.)  Then the it could be invisible to the user.

**Comment by softins** [comment]
I did try something like that earlier, but found that it wasn't easy to pass the result of the test back up from the `CSocket`, given that the client and server sockets are not newed, but instantiated as members of `CClient` or `CServer`.

I could perhaps create a static method of `CSocket` that just tries to create an AF_INET6 socket and returns a success indication, and call that from `main.cpp`

We also need to consider the implications for:

* A client on a computer that is IPv6-capable, but connected via an internet connection that does not support IPv6
* A server that supports IPv6 and has a routable IPv6 address, but the operator does not want their Jamulus server to listen on IPv6, for whatever reason.

**Comment by pljones** [comment]
You've given reasons to switch IPv6 off.  Maybe the command line parameter should do that?

I'm thinking of other software with internet connectivity -- I don't have to tell my web browser or email client whether or not to support IPv6, for example.

**Comment by ann0see** [comment]
> A client on a computer that is IPv6-capable, but connected via an internet connection that does not support IPv6

Best case if host name is given: try to request an A record via DNS. Use ipv4 as fallback. But does Jamulus need to handle that? Isn’t that an OS functionality? Not only Jamulus will have this problem?!

If I‘m giving an ipv6 address it should fail and not try to get a connection over ipv4.

How does Firefox/Chrome/whatever handle this?

**Comment by ann0see** [comment]
> A server that supports IPv6 and has a routable IPv6 address, but the operator does not want their Jamulus server to listen on IPv6, for whatever reason.

Use the bind IP option and only provide an ipv6/ipv4 address?

**Comment by atsampson** [comment]
> How does Firefox/Chrome/whatever handle this?

[Happy Eyeballs](https://datatracker.ietf.org/doc/html/rfc8305).

Firefox does have a configuration option to disable IPv6 entirely to speed things up a bit if you know you're on a network without external IPv6 connectivity (it's still not that unusual to have local IPv6 but not routable to the world).

**Comment by softins** [comment]
> > How does Firefox/Chrome/whatever handle this?
> 
> [Happy Eyeballs](https://datatracker.ietf.org/doc/html/rfc8305).

Certainly very interesting, but quite complex, and I feel it's likely to be overkill for Jamulus. It's primarily concerned with resolving DNS names using both AAAA and A records, and then trying to connect with the returned addresses, preferring IPv6. In Jamulus, an operator of a private server using IPv6 will likely give his invitees a literal address or have a specific AAAA record for it. And once we extend to directory servers, they will return to the client literal IPv6 (if any) and IPv4 addresses, not DNS names.

**Comment by softins** [comment]
> You've given reasons to switch IPv6 off. Maybe the command line parameter should do that?

Maybe, but I think the number of people with an IPv6-capable computer but an IPv4-only internet connection would be a lot larger than the number of people who specifically want to use IPv6. So we would be requiring the much larger number of people to do something different.

**Comment by ann0see** [comment]
> In Jamulus, an operator of a private server using IPv6 will likely give his invitees a literal address or have a specific AAAA record for it. 

What makes you sure about that?

From a UX point of view, I think just giving one address (with AAAA and A records) is the preferable outcome. Maybe - but for debugging purposes only - the admin sends over an IP to force ipv6/ipv4. That’s just my opinion…

For directories we‘re skipping DNS? 

Can we send ping messages over IPv6 and IPv4, and constantly calculate the time difference between both as soon as the connection window is open? The client will then know which one to choose before connecting. 

Issues:
* Worst case: double the ping traffic
* Complexity on the client side
* Constant calculation might not be necessary. We might stop after a few ping messages.

Pro:
* Probably similar to "Happy eyeballs"
* User doesn‘t need to decide in normal mode, but could still use the direct IP to connect to the server

**Comment by ann0see** [comment]
So -6 is a client and server option?

**Comment by softins** [comment]
> So -6 is a client and server option?

Yes.

**Comment by softins** [comment]
> > In Jamulus, an operator of a private server using IPv6 will likely give his invitees a literal address or have a specific AAAA record for it.
> 
> What makes you sure about that?

I'm not sure, it just seems likely to me. Most people I know with private servers haven;t set up a DNS name and just give their friends the IP.

> For directories we‘re skipping DNS?

We can use a DNS name to find a directory, but when a server registers with a directory, it does so just by IP (the directory stores the IP it registered from). The server then tells a requesting client just the IP, not a DNS name, plus other relevant information (private IP, port number, etc).

> Can we send ping messages over IPv6 and IPv4, and constantly calculate the time difference between both as soon as the connection window is open? The client will then know which one to choose before connecting.

That might well be a useful enhancement for the future, but I don't think we should hold up the basic feature waiting for that particular lily to be gilded.

**Comment by ann0see** [comment]
Ok. So what really needs to be tested to get this merged?

IPv6/IPv4 over a dual stack connection seems to work.

**Comment by softins** [comment]
> Ok. So what really needs to be tested to get this merged?

Well I've tested it quite a bit in my environment, on Linux, Windows and Mac, and it's running on my server as mentioned above. I think it's good to go, and that it is beneficial to have `-6` to explicitly turn it on for now. Then others who want to try out IPv6 could do so, without it having unexpected impacts on those who don't.

So it just needs two approvals (which can be tricky), and it could be merged. I would then start thinking about Directory Servers.

> IPv6/IPv4 over a dual stack connection seems to work.

Cool!

**Comment by pljones** [comment]
> it is beneficial to have -6 to explicitly turn it on for now.

Agreed for the "experimental support" version.

File: src/client.cpp
Comment (pljones) [reply src/client.cpp:34]: I'm not going to block it for that, either.  (It's just on my list of things to get rid of.)

File: src/socket.cpp
Comment (pljones) [inline src/socket.cpp:86]: A socket listening on an IPv4 address is only listening for IPv4 traffic.  But the same _interface_ could have a separate socket listening on an IPv6 address, I believe.  The _interface_ has the MAC address (or equivalent).  The protocol has the IPv4 or IPv6 address that goes with the socket layer in use.

At least, that's my understanding.

File: src/socket.cpp
Comment (pljones) [inline src/socket.cpp:235]: Can this go in a header somewhere as a type, so it can be re-used more easily?

Comment (softins) [reply src/socket.cpp:235]: I guess, although that makes it more opaque when reading `socket.cpp`.

File: src/socket.cpp
Comment (softins) [reply src/socket.cpp:81]: > A socket listening on an IPv4 address is only listening for IPv4 traffic. But the same _interface_ could have a separate socket listening on an IPv6 address, I believe.

Yes it could, but Jamulus only expects to listen on one socket. So I think that makes it unable to bind to a specific interface address and remain dual-stack. I don't think it's a problem, unless someone flags up a scenario where it is. The default of binding to `::` will allow it to receive both IPv4 and IPv6 traffic without needing two sockets.

File: src/socket.cpp
Comment (pljones) [reply src/socket.cpp]: You could argue that for every definition that isn't in line.

File: src/socket.cpp
Comment (pljones) [reply src/socket.cpp:81]: > Yes it could, but Jamulus only expects to listen on one socket.
> The default of binding to :: will allow it to receive both IPv4 and IPv6 traffic without needing two sockets.

How does an IPv4 client know what address to use?  Does the IPv6 socket "just know" it's got two addresses really?

**Review by pljones — APPROVED** [review]

File: src/util.cpp
Comment (ann0see) [inline src/util.cpp:735]: Do we give an error message to the user somewhere else?

**Review by ann0see — COMMENTED** [review]
Seems simple. Nevertheless some comments:
1. You're removing the chinese .qm file? Any reason for that?
2. See comments

**Comment by ann0see** [comment]
Since it's not enabled by default, I think we can tag it for 3.8.1.

**Comment by softins** [comment]
> Seems simple. Nevertheless some comments:
> 
> 1. You're removing the chinese .qm file? Any reason for that?

No, must be an accident. I'll correct that before merging.

> 2. See comments

Thanks, will respond individually.

> 3. As it works on my ipv6 connection (and you tested it), I think the functionality is ok.

Cool.

File: src/serverlist.cpp
Comment (softins) [reply src/serverlist.cpp:136]: I don't agree with this one. When IP addresses are shown (e.g. by `ip addr` or `host`), they are not quoted, as they are values in their own right, not just strings. It would be like displaying the numeric port number as `'22124'` instead of `22124`.

File: src/serverlist.cpp
Comment (softins) [reply src/serverlist.cpp:127]: As above

File: src/util.cpp
Comment (softins) [reply src/util.cpp:735]: At the moment, with all versions, if an invalid hostname or IP address is entered into the Connect window, the connection fails silently. It would be good to fix that more generally, but I think that should be separate from this PR. Maybe raise it as a specific issue?

File: src/socket.cpp
Comment (softins) [reply src/socket.cpp:63]: I can't see any reference to active vs passive voice in that page, although I know some other wider contexts promote a preference for active voice. In this case, I have a slight preference for the shorter message.

File: src/socket.cpp
Comment (softins) [reply src/socket.cpp:167]: This was just a code formatting change; the original wording was Volker's. Happy to change it if you prefer.

File: src/socket.cpp
Comment (softins) [reply src/socket.cpp:81]: The IPv4 client will connect in the same way as at present, by being told the IP address, either explicitly, or from a hostname lookup, or by a directory server.

Binding an `AF_INET6` socket to `::` (declared as `in6addr_any`) is the IPv6 equivalent of binding an `AF_INET` socket to `0.0.0.0` (`INADDR_ANY`), which means it will accept any incoming packet that arrives on any interface. If the socket has been `setsockopt`ed to dual stack (as we do), it will accept both IPv4 and IPv6 packets.

An IPv4 address mapped to IPv6 uses `0000:0000:0000:0000:0000:ffff` as the first 96 bits, which denotes that the last 32 bits contain an IPv4 address in network byte order. Often expressed like `::ffff:192.168.1.1`. This mapping is used in `CSocket::SendPacket()` and checked for in `CSocket::OnDataReceived()`.

File: src/serverlist.cpp
Comment (ann0see) [reply src/serverlist.cpp:136]: > It would be like displaying the numeric port number as '22124' instead of 22124.

Yes. It’s just cosmetics/personal preference. Since we have spaces before and after the IP and they all follow a standard probably the quotes are not needed.

File: src/socket.cpp
Comment (ann0see) [reply src/socket.cpp:63]: I think @gilgongo mentioned it somewhere.

File: src/socket.cpp
Comment (ann0see) [reply src/socket.cpp:167]: Yes. Or part of https://github.com/jamulussoftware/jamulus/pull/1889

File: src/util.cpp
Comment (ann0see) [reply src/util.cpp:735]: Done #1945

File: src/main.cpp
Comment (pljones) [reply src/main.cpp:958]: I don't think this adds anything at all.  It still requires further explanation as to what "direct connection to client/server" means, so it's useless verbosity.

Comment (ann0see) [reply src/main.cpp:958]: But do we agree that the original proposal needs more clarification?

File: src/socket.cpp
Comment (gilgongo) [reply src/socket.cpp:63]: Maybe I halluciated it then - I recall @mulyaj mentioning it.

**Comment by ann0see** [comment]
As soon as the Chinese qm file is added, I think it’s ready to merge.

File: src/main.cpp
Comment (pljones) [reply src/main.cpp:958]: I'm happy with the wording Tony used.  It's consise and says what it does.

File: src/socket.cpp
Comment (mulyaj) [reply src/socket.cpp:63]: Yeah, I've mentioned a preference for active voice. Reading this at first glance, the subject is vague: 

"IPv6 requested..." - who requested?

Then, if the subject is vague, the guideline for concise and specific is broken.

Comment (pljones) [reply src/socket.cpp:63]: I'd be happy with "IPv6 is not available."  It's issued as a response to the user making a request to use an address.  It doesn't need to tell the user they used an address, just why it hasn't worked.

File: src/main.cpp
Comment (ann0see) [reply src/main.cpp:958]: I still think it's not clear. Probably the question is:
*Does the user need to know that connections to public servers is still IPv4 only?*
If we want to be specific: Yes. If we think it confuses the user: No.
Maybe it's worth mentioning it at least in the release announcement.

**Review by ann0see — APPROVED** [review]
Since it works and only one critical step is missing (.qm file), I‘ll approve it.

File: src/main.cpp
Comment (pljones) [reply src/main.cpp:958]: > Does the user need to know that connections to public servers is still IPv4 only?

Not "public", use the word "registered".  You can register with a directory no one knows about.  That's not particularly public.

To operate the software, no.  They can just select a listed entry from the directory and the software will use the returned IPv4 address.

If you want to use IPv6, you need to:
a) pass "-6" on the command line, thus indicating you understand how to operate with IPv6 addresses, so need no further help
b) enter an IPv6 name or address in the connection dialogue "Server Address", by hand, so you know that you've chosen a specific server rather than a registered one.

As I say, the help text need not say anything else.  This is not the final form for the feature.  It's intended to be somewhat experimental and the documentation should make that clear.  Those who choose to use it should thus be given the credit for understanding what they've chosen to do.

**Comment by softins** [comment]
I have squashed all the commits except for the first one by @jardous. I'll wait until the checks are complete (looks like they may need to be rerun) and merge tomorrow.

**Comment by ann0see** [comment]
I think you should use @jardous E-Mail or jardous@users.noreply.github.com to link his GitHub account to the first commit. This would add him to the contributor entry on GitHub.

**Comment by softins** [comment]
> I think you should use @jardous E-Mail or [jardous@users.noreply.github.com](mailto:jardous@users.noreply.github.com) to link his GitHub account to the first commit. This would add him to the contributor entry on GitHub.

Not exactly sure how to do that.

**Comment by pljones** [comment]
> 
> 
> > I think you should use @jardous E-Mail or [jardous@users.noreply.github.com](mailto:jardous@users.noreply.github.com) to link his GitHub account to the first commit. This would add him to the contributor entry on GitHub.
> 
> Not exactly sure how to do that.

To set the values:
```
git commit --author="Jiri Popek <jardous@users.noreply.github.com>" --date="2021-02-18T12:34:56" ...
```
You'd need to do that during either a `git rebase -i` or manually `git reset --hard master` and reapplying the commits.

**Comment by softins** [comment]
> I think you should use @jardous E-Mail or [jardous@users.noreply.github.com](mailto:jardous@users.noreply.github.com) to link his GitHub account to the first commit. This would add him to the contributor entry on GitHub.

OK, that seems to have worked. Just waiting for the Checks to run and then I'll merge.

**Comment by softins** [comment]
Well although Android and Windows failed this time, they have previously worked with identical code (I diffed the before and after of the rebases to be sure), so I'll go ahead and merge anyway.

**Comment by jardous** [comment]
hi. Feel free to use my real email: ***@***.***
J.

**Comment by softins** [comment]
> hi. Feel free to use my real email: ***@***.*** J.

I already used your generic Github address jardous@users.noreply.github.com, which was enough to link the rebased version of your original commit to your Github ID. Thanks!

**Pull request merged**
